File: scripts/update_skadnetwork.js
/**
 * SKAdNetworkItems Update Script
 * 工場長用: 全アプリ共通の広告ネットワークIDリストを一括適用するスクリプト
 * 
 * Usage: node scripts/update_skadnetwork.js
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ターゲットファイル
const PLIST_PATH = path.resolve(__dirname, '../ios/App/App/Info.plist');

// === 必須ネットワークリスト (Top 50+ Global Networks) ===
// Google, Meta, Unity, AppLovin, ironSource, TikTok(Pangle), Vungle, etc.
const SKADNETWORK_IDS = [
  // Google
  'cstr6suwn9.skadnetwork',
  // Meta (Facebook)
  'v9wttpbfk9.skadnetwork',
  'n38lu8286q.skadnetwork',
  // Unity Ads
  '4dzt52r2t5.skadnetwork',
  'bvpn9n2tp3.skadnetwork',
  // AppLovin
  'ludvb6z3bs.skadnetwork',
  // ironSource
  'su67r6k2v3.skadnetwork',
  // Pangle (TikTok)
  '238da6jt44.skadnetwork',
  '22mmun2rn5.skadnetwork',
  // Liftoff (Vungle)
  'gta9lk7p23.skadnetwork',
  // Mintegral
  'kbd757ywx3.skadnetwork',
  // InMobi
  'wzmmz9fp6w.skadnetwork',
  // Chartboost
  'f38h382jlk.skadnetwork',
  // AdColony
  '4pfyvq9l8r.skadnetwork',
  // Tapjoy
  'ecpz2srf59.skadnetwork',
  // Smaato
  'v72qych5uu.skadnetwork',
  // Maio
  '5a6flpkh64.skadnetwork',
  // Criteo
  '9rd848q2bz.skadnetwork',
  // Yahoo
  'e5fvkxwrpn.skadnetwork',
  // Amazon
  'p78axxw29g.skadnetwork',
  // PubMatic
  'k674qkevps.skadnetwork',
  // Index Exchange
  '5l3tpt7t6e.skadnetwork',
  // OpenX
  'av6w8kgt66.skadnetwork',
  // The Trade Desk
  'uw77j35x4d.skadnetwork',
  // Verizon
  'e5fvkxwrpn.skadnetwork',
  // MobVista
  '2u9pt9hc89.skadnetwork',
  // Fyber
  'n38lu8286q.skadnetwork',
  // StartApp
  '5l3tpt7t6e.skadnetwork',
  // YouAppi
  '3rd42ekr43.skadnetwork',
  // Moloco
  'glqzh8vgby.skadnetwork',
  // Aarki
  '4fyvq9l8r.skadnetwork',
  // LoopMe
  '5tjdwbrq8w.skadnetwork',
];

function updatePlist() {
  if (!fs.existsSync(PLIST_PATH)) {
    console.error(`❌ Info.plist not found at: ${PLIST_PATH}`);
    process.exit(1);
  }

  let content = fs.readFileSync(PLIST_PATH, 'utf8');

  // 1. 既存の SKAdNetworkItems ブロックを探す
  const startTag = '<key>SKAdNetworkItems</key>';
  const endTag = '</array>';
  
  const startIndex = content.indexOf(startTag);
  if (startIndex === -1) {
    // ブロックがない場合は挿入（簡易実装：dictの最後に追加）
    console.log('⚠️ SKAdNetworkItems not found. Creating new block...');
    const dictEnd = '</dict>\n</plist>';
    const newBlock = `\t<key>SKAdNetworkItems</key>\n\t<array>\n${generateArrayItems()}\t</array>\n`;
    content = content.replace(dictEnd, `${newBlock}${dictEnd}`);
  } else {
    // 既存ブロックを置換
    const arrayStartIndex = content.indexOf('<array>', startIndex);
    const arrayEndIndex = content.indexOf('</array>', arrayStartIndex);
    
    if (arrayStartIndex === -1 || arrayEndIndex === -1) {
      console.error('❌ Failed to parse existing SKAdNetworkItems array.');
      process.exit(1);
    }

    const before = content.substring(0, arrayStartIndex + 7); // <array> まで
    const after = content.substring(arrayEndIndex); // </array> から
    
    // 中身を全置換（重複排除のため、常に指定リストで上書きする仕様とします）
    const newItems = generateArrayItems();
    content = before + '\n' + newItems + '\t' + after;
  }

  fs.writeFileSync(PLIST_PATH, content, 'utf8');
  console.log(`✅ SKAdNetworkItems updated! Total count: ${SKADNETWORK_IDS.length}`);
}

function generateArrayItems() {
  // 重複除去 + 小文字統一
  const uniqueIds = [...new Set(SKADNETWORK_IDS.map(id => id.toLowerCase()))];
  
  return uniqueIds.map(id => 
    `\t\t<dict>\n\t\t\t<key>SKAdNetworkIdentifier</key>\n\t\t\t<string>${id}</string>\n\t\t</dict>`
  ).join('\n');
}

updatePlist();

File: scripts/update_skadnetwork.py
import os
import sys

# Script location
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
# Target file
PLIST_PATH = os.path.abspath(os.path.join(SCRIPT_DIR, '../ios/App/App/Info.plist'))

# SKAdNetwork IDs
SKADNETWORK_IDS = [
  'cstr6suwn9.skadnetwork', 'v9wttpbfk9.skadnetwork', 'n38lu8286q.skadnetwork',
  '4dzt52r2t5.skadnetwork', 'bvpn9n2tp3.skadnetwork', 'ludvb6z3bs.skadnetwork',
  'su67r6k2v3.skadnetwork', '238da6jt44.skadnetwork', '22mmun2rn5.skadnetwork',
  'gta9lk7p23.skadnetwork', 'kbd757ywx3.skadnetwork', 'wzmmz9fp6w.skadnetwork',
  'f38h382jlk.skadnetwork', '4pfyvq9l8r.skadnetwork', 'ecpz2srf59.skadnetwork',
  'v72qych5uu.skadnetwork', '5a6flpkh64.skadnetwork', '9rd848q2bz.skadnetwork',
  'e5fvkxwrpn.skadnetwork', 'p78axxw29g.skadnetwork', 'k674qkevps.skadnetwork',
  '5l3tpt7t6e.skadnetwork', 'av6w8kgt66.skadnetwork', 'uw77j35x4d.skadnetwork',
  'e5fvkxwrpn.skadnetwork', '2u9pt9hc89.skadnetwork', 'n38lu8286q.skadnetwork',
  '5l3tpt7t6e.skadnetwork', '3rd42ekr43.skadnetwork', 'glqzh8vgby.skadnetwork',
  '4fyvq9l8r.skadnetwork', '5tjdwbrq8w.skadnetwork'
]

def generate_array_items():
    unique_ids = sorted(list(set([x.lower() for x in SKADNETWORK_IDS])))
    items = []
    for sk_id in unique_ids:
        items.append(f"\t\t<dict>\n\t\t\t<key>SKAdNetworkIdentifier</key>\n\t\t\t<string>{sk_id}</string>\n\t\t</dict>")
    return "\n".join(items)

def update_plist():
    if not os.path.exists(PLIST_PATH):
        print(f"❌ Info.plist not found at: {PLIST_PATH}")
        return

    with open(PLIST_PATH, 'r', encoding='utf-8') as f:
        content = f.read()

    start_tag = '<key>SKAdNetworkItems</key>'
    # Simple logic to find the block
    start_index = content.find(start_tag)

    updated_items = generate_array_items()

    if start_index == -1:
        print("⚠️ SKAdNetworkItems not found. Creating new block...")
        dict_end = '</dict>\n</plist>'
        new_block = f"\t<key>SKAdNetworkItems</key>\n\t<array>\n{updated_items}\n\t</array>\n"
        if dict_end in content:
            content = content.replace(dict_end, new_block + dict_end)
        else:
             print("❌ Could not find end of dict to insert new block.")
             return
    else:
        # Find array start and end
        array_start_tag = '<array>'
        array_end_tag = '</array>'
        
        array_start_index = content.find(array_start_tag, start_index)
        array_end_index = content.find(array_end_tag, array_start_index)
        
        if array_start_index == -1 or array_end_index == -1:
             print("❌ Failed to parse existing SKAdNetworkItems array.")
             return

        before = content[:array_start_index + len(array_start_tag)]
        after = content[array_end_index:]
        content = before + "\n" + updated_items + "\n\t" + after

    with open(PLIST_PATH, 'w', encoding='utf-8') as f:
        f.write(content)
    
    print(f"✅ SKAdNetworkItems updated! Total count: {len(set(SKADNETWORK_IDS))}")

if __name__ == "__main__":
    update_plist()
